File: frontend/src/api/specialist.ts
/** Control-centre endpoints: the specialist's decisions (persisted) and the assistant proxy. */
import { useQuery } from "@tanstack/react-query";
import type { AssistantRequest, SpecialistDecisionCreate } from "./generated";
import { buildPath, request } from "./client";
import {
  bool,
  isoDate,
  isoDateTime,
  literal,
  nullable,
  num,
  object,
  str,
} from "./schema";
import { pageSchema } from "./types";

export const specialistDecisionSchema = object({
  id: num,
  created_at: isoDateTime,
  origin: isoDate,
  run_id: nullable(str),
  sim_day: num,
  subject_kind: literal("alert", "patient"),
  subject_id: str,
  region_code: nullable(str),
  org_code: nullable(str),
  profile_code: nullable(str),
  action: literal("accept", "decline", "clarify", "confirm", "postpone"),
  comment: nullable(str),
  actor: nullable(str),
  idempotency_key: nullable(str),
  api_key_label: nullable(str),
});
export const assistantStatusSchema = object({
  configured: bool,
  provider: nullable(str),
  model: nullable(str),
});
export const assistantReplySchema = object({
  text: str,
  provider: str,
  model: str,
});

export const specialistApi = {
  decisions: (origin: string, runId: string) =>
    request(
      pageSchema(specialistDecisionSchema),
      buildPath("/specialist-decisions", {
        origin,
        run_id: runId,
        limit: 500,
        offset: 0,
      }),
    ),
  createDecision: (payload: SpecialistDecisionCreate) =>
    request(specialistDecisionSchema, "/specialist-decisions", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    }),
  assistantStatus: () => request(assistantStatusSchema, "/assistant/status"),
  ask: (payload: AssistantRequest) =>
    request(assistantReplySchema, "/assistant", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    }),
};

export const useSpecialistDecisions = (
  origin: string | null,
  runId: string | null,
) =>
  useQuery({
    queryKey: ["specialist-decisions", origin, runId],
    queryFn: () => specialistApi.decisions(origin ?? "", runId ?? ""),
    enabled: Boolean(origin && runId),
    staleTime: 60_000,
    retry: false,
  });

export const useAssistantStatus = () =>
  useQuery({
    queryKey: ["assistant", "status"],
    queryFn: specialistApi.assistantStatus,
    staleTime: 5 * 60_000,
    retry: false,
  });

File: backend/app/schemas/specialist.py
"""Control-centre contracts: the specialist's decisions and the assistant proxy (docs/api.md §6)."""

import datetime as dt
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

SubjectKind = Literal["alert", "patient"]
SpecialistAction = Literal["accept", "decline", "clarify", "confirm", "postpone"]


class SpecialistDecisionCreate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    origin: dt.date = Field(description="publication origin the demo runs on")
    run_id: str | None = Field(
        default=None,
        max_length=64,
        pattern=r"^[A-Za-z0-9._:-]+$",
        description="simulation run the decision belongs to; a restart of the simulation starts a new run, and "
        "the control centre replays only the decisions of its current run",
    )
    sim_day: int = Field(ge=0, le=60, description="day of the simulation when the decision was taken (0 = origin)")
    subject_kind: SubjectKind
    subject_id: str = Field(min_length=1, max_length=64, description="signal id or synthetic referral id")
    region_code: str | None = Field(default=None, max_length=4)
    org_code: str | None = Field(default=None, max_length=8)
    profile_code: str | None = Field(default=None, max_length=8)
    action: SpecialistAction = Field(
        description="alerts: accept | decline | clarify; admission requests: confirm | decline | postpone"
    )
    comment: str | None = Field(default=None, max_length=4000)
    actor: str | None = Field(default=None, max_length=200)
    idempotency_key: str | None = Field(
        default=None,
        min_length=8,
        max_length=128,
        pattern=r"^[A-Za-z0-9._:-]+$",
        description="client-generated per submission; resending the same key returns the stored row with 200",
    )


class SpecialistDecision(SpecialistDecisionCreate):
    id: int
    created_at: dt.datetime
    api_key_label: str | None = Field(description="label of the API key that submitted the decision")


class AssistantStatus(BaseModel):
    configured: bool = Field(description="an external model is configured on the server")
    provider: str | None = Field(description="groq | openrouter | gemini | openai-compatible")
    model: str | None


class AssistantSubject(BaseModel):
    kind: SubjectKind
    id: str = Field(max_length=64)
    hospital: str = Field(max_length=300)


class AssistantRequest(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    lang: Literal["ru", "kk"] = "ru"
    question: str = Field(min_length=1, max_length=2000)
    facts: list[str] = Field(default_factory=list, max_length=40)
    explanation: list[str] = Field(default_factory=list, max_length=12)
    subject: AssistantSubject


class AssistantReply(BaseModel):
    text: str
    provider: str
    model: str
